text/0000-no-more-extern-crate.md

```diff
@@ -0,0 +1,136 @@
+the `extern crate` declarations.
+
+This duplication causes unnecessary effort and results in one more opportunity
+for mistakes when working with conditionally-enabled dependencies.
+Allowing the omission of the redundant `extern crate` syntax contributes to the
+roadmap goals of
+[improving Rust's ergonomics](https://github.com/rust-lang/rust-roadmap/issues/17)
+and
+[providing easy access to high-quality crates.](https://github.com/rust-lang/rust-roadmap/issues/9)
+
+# Guide-Level Explanation
+[guide]: #guide
+
+When you add a dependency to your `Cargo.toml`, it is immediately usable within
+the source of your crate:
+
+```toml
+# Cargo.toml:
+name = "my_crate"
+version = "0.1.0"
+authors = ["Me" <me@mail.com>]
+
+[dependencies]
+rand = "0.3"
+```
+
+```rust
+// src/main.rs:
+
+fn main() {
+    println!"A random character: {}", rand::random::<char>());
+}
+```
+
+# Reference-Level Explanation
+[reference]: #reference
+
+External crates can be passed to the rust compiler using the
+`--extern CRATE_NAME=PATH` flag.
+For example, `cargo build`-ing a crate `my_crate` with a dependency on `rand`
+results in a call to rustc that looks something like
+`rustc --crate-name mycrate src/main.rs --extern rand=/path/to/librand.rlib ...`.
+
+When an external crate is specified this way,
+the crate will automatically brought into scope as if an
+`extern crate name_of_crate;`
+declaration had been added to the current crate root.
+This behavior won't occur when including a library using the `-l`
+or `-L` flags.
+
+We will continue to support the current `extern crate` syntax,
+both for backwards compatibility and to enable users who want to use manual
+`extern crate` in order to have more fine grained control-- say, if they wanted
+to import an external crate only inside an inner module.
+No automatic import will occur if an `extern crate` declaration for the same
+external dependency appears anywhere within the crate.
+For example, if `rand = "0.3"` is listed as a dependency in Cargo.toml
+and `extern crate rand;` appears somewhere in the crate being compiled,
+then no implicit `extern crate rand;` will be added.
+If Cargo.toml were to also list another dependency, `log = "0.3"`, and no
+`extern crate log;` appears in the crate being compiled,
+then an `extern crate log;` would be implicitly added.
+
+Additionally, items such as modules, types, or functions that conflict with
+the names of implicitly imported crates will cause the implicit `extern crate`
+declaration to be removed.
+Note that this is different from the current behavior of the
+implicitly-imported `std` module.
+Currently, creating a root-level item named `std` results in a name conflict
+error. For consistency with other crates, this error will be removed.
+Creating a root-level item named `std` will prevent `std` from being included,
+and will trigger a warning.
+
+It will still be necessary to use the `extern crate` syntax when using
```

Comment:
Yes, it's definitely annoying. Since we do have a plan to migrate away from `#[macro_use]` in the relatively near future, I think it's livable, but the overlap period of "`extern crate` just for macros" is frustrating.